ang/test/civicase/ActivityActions--tags.spec.js

```diff
@@ -0,0 +1,362 @@
+          expect($.fn.dialog).toHaveBeenCalledWith('close');
+        });
+      });
+    });
+
+    describe('Remove Tags to Activities bulk action', function () {
+      var modalOpenCall;
+
+      beforeEach(function () {
+        crmApiMock.and.returnValue($q.resolve([{ values: TagsMockData.get() }]));
+        TagsActivityAction.manageTags('remove', activitiesMockData.get());
+        $rootScope.$digest();
+        modalOpenCall = dialogServiceMock.open.calls.mostRecent().args;
+      });
+
+      it('fetches the tags from the api endpoint', function () {
+        expect(crmApiMock).toHaveBeenCalledWith([['Tag', 'get', {
+          'sequential': 1,
+          'used_for': { 'LIKE': '%civicrm_activity%' }
+        }]]);
+      });
+
+      it('sets the title of the modal as "Tag Activities (Remove)"', function () {
+        expect(modalOpenCall[3].title).toBe('Tag Activities (Remove)');
+      });
+
+      it('sets the text of the save button of the modal as "Remove tags from Activities"', function () {
+        expect(modalOpenCall[3].buttons[0].text).toBe('Remove tags from Activities');
+      });
+
+      it('does not have any selected generic tags initially', function () {
+        expect(modalOpenCall[2].selectedGenericTags.length).toBe(0);
+      });
+
+      describe('when generic tags are toggled', function () {
+        describe('when a checkbox is enabled', function () {
+          beforeEach(function () {
+            modalOpenCall[2].toggleGenericTags(modalOpenCall[2], TagsMockData.get(0).id);
+          });
+
+          it('add the clicked tag as a selected tag', function () {
+            expect(modalOpenCall[2].selectedGenericTags).toEqual([TagsMockData.get(0).id]);
+          });
+        });
+
+        describe('when a checkbox is disabled after being enabled', function () {
+          beforeEach(function () {
+            modalOpenCall[2].toggleGenericTags(modalOpenCall[2], TagsMockData.get(0).id);
+            modalOpenCall[2].toggleGenericTags(modalOpenCall[2], TagsMockData.get(0).id);
+          });
+
+          it('removes the clicked tag as a selected tag', function () {
+            expect(modalOpenCall[2].selectedGenericTags).toEqual([]);
+          });
+        });
+      });
+
+      describe('generic tags', function () {
+        describe('generic tags does not include tagsets', function () {
+          var genericTagsHasTagSets;
+
+          beforeEach(function () {
+            genericTagsHasTagSets = false;
+
+            modalOpenCall[2].genericTags.forEach(function (tag) {
+              if (tag.is_tagset !== '0') {
+                genericTagsHasTagSets = true;
+              }
+            });
+          });
+
+          it('does not include tagsets as generic tags', function () {
+            expect(genericTagsHasTagSets).toBe(false);
+          });
+        });
+
+        describe('child tags are indented in the UI', function () {
+          var tagWhichHaveOneLevelOfParent, tagWhichHaveTwoLevelofParent;
```

Comment:
Recommendation - 
Can we rename the vars with smaller names `tagWithOneLevelOfParent` and `tagWithTwoLevelofParent`

We can keep them as it as well. Just a recommendation